storage/range.go

```diff
@@ -822,119 +858,64 @@ func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNode p
 		// since reads are served locally by the lease holder without going
 		// through Raft, a read which was not taken into account may have been
 		// served. Hence, we must retry at the current leader.
-		//
-		// It's crucial that we don't update the response cache for the error
-		// returned below since the request is going to be retried with the
-		// same ClientCmdID and would get the distributed sender stuck in an
-		// infinite loop, retrieving a stale NotLeaderError over and over
-		// again, even when proposing at the correct replica.
-		return r.newNotLeaderError(lease, originNode)
-	}
-
-	// Anything happening from now on needs to enter the response cache.
-	defer func() {
-		// TODO(tamird,tschottdorf): according to #1400 we intend to set the reply
-		// header's error as late as possible and in a central location. Range
-		// commands still write to the header directly, but once they don't this
-		// could be the authoritative location that sets the reply error for any-
-		// thing that makes it into Raft. Note that we must set this prior to
-		// signaling cmd.done below, or the waiting RPC handler might proceed
-		// before we've updated its reply.
-		//
-		// It is important that the error is set before the reply is saved into
-		// the response cache.
-		reply.Header().SetGoError(rErr)
-
-		if proto.IsWrite(args) {
-			// No matter the result, add result to the response cache if this
-			// is a write method. This must be done as part of the execution of
-			// raft commands so that every replica maintains the same responses
-			// to continue request idempotence, even if leadership changes.
-			if err := r.respCache.PutResponse(args.Header().CmdID, reply); err != nil {
-				rErr = newReplicaCorruptionError(
-					util.Errorf("could not put to response cache"), err, rErr)
-				return
-			}
-		}
-	}()
-
-	header := args.Header()
+		return batch, r.newNotLeaderError(lease, originNode)
+	}
 
 	// Check the response cache to ensure idempotency.
 	if proto.IsWrite(args) {
-		if ok, err := r.respCache.GetResponse(header.CmdID, reply); ok && err == nil {
+		if ok, err := r.respCache.GetResponse(batch, args.Header().CmdID, reply); ok && err == nil {
 			if log.V(1) {
 				log.Infoc(ctx, "found response cache entry for %+v", args.Header().CmdID)
 			}
 			// We successfully read from the response cache, so return whatever error
 			// was present in the cached entry (if any).
-			return reply.Header().GoError()
+			return batch, reply.Header().GoError()
 		} else if ok && err != nil {
```

Comment:
why are we checking `ok` here? If it returns an error either way, we should sense corruption.